### ISSUE TITLE
shared cache search and searchByMime don't scale well

### ISSUE THREAD

**Issue by butonic** [owncloud/core#10077]
The current implementation does not scale because `searchByMime()` as well as `search()` uses dozens of sql queries when walking the directory tree: getFolderContents first looks up the source cache via the shares table and that then looks up the files in the filecache. That is at least two queries per folder from what I can tell. Have a look at the following query log: http://pastebin.com/6V0M2NAg

And I already stripped it of completely unneeded duplicate queries for user group and what not. But that is yet another issue.

What I propose is having the current filecache and share tables and an additional relation table called 'oc_shares_files' having the fileid and the shareid as foreign keys:

``` SQL
CREATE TABLE `oc_shares_files` (`shareid` int NOT NULL , `fileid` int NOT NULL , PRIMARY KEY (`shareid`, `fileid`));

-- We can then join the share table via the relations to search files
SELECT "sf"."shareid", "fc"."fileid"
FROM "oc_filecache" "fc"
JOIN "oc_share_files" "sf" ON "sf"."fileid"="fc"."fileid"
JOIN "oc_share" "s" ON "sf"."shareid"="s"."id"
 WHERE "fc"."name" LIKE ?
AND "s"."share_with" = ?
```

pass in the pattern and the user searching through shared files. The query could be optimized to also show files owned by the user.

Maintaining the relations is not that hard:

``` SQL
-- add all children i nthe same storage when a new share is created: params are shareid, fileid of the shared file and path of the shared file with % appended,
INSERT INTO  `oc_shares_files` (`shareid`,`fileid`)
  SELECT ?, `fc`.`fileid` 
  FROM `oc_filecache` `fc` 
    JOIN `oc_filecache` `fc2` ON `fc`.`fileid` = `fc`.`fileid`
  WHERE `fc`.`storage` = `fc2`.`storage`
    AND `fc2`.`fileid` = ?
    AND `fc`.`path` LIKE ?
```

All this happens inside the database. A quick test on an oracle db that had to insert 7736 rows took less than a second. Honestly, this is what RDBMS are made for. On my productive mysql machine I did:

``` SQL
SELECT count(*) FROM `oc_filecache` WHERE `path` LIKE 'files%';
+----------+
| count(*) |
+----------+
|    21750 |
+----------+
1 row in set (0.19 sec)

INSERT INTO `oc_shares_files` (`shareid`,`fileid`) SELECT 1, `fc`.`fileid`    FROM `oc_filecache` `fc`      JOIN `oc_filecache` `fc2` ON `fc`.`fileid` = `fc`.`fileid`   WHERE `fc`.`storage` = `fc2`.`storage`     AND `fc2`.`fileid` = 2938     AND `fc`.`path` LIKE 'files%';
Query OK, 17261 rows affected (0.04 sec)
Records: 17261  Duplicates: 0  Warnings: 0
```

so ... IMHO performance is clearly not an issue here.

Other operations can now also be offloaded to the database:

``` SQL
-- on a create hook add the file to the shares_files relation by looking up the parent in the share_files table
INSERT INTO `oc_shares_files` (`shareid`,`fileid`) SELECT `sf`.`shareid`, `fc`.`fileid` FROM `oc_filecache` `fc` JOIN `oc_share_files` `sf` ON `sf`.`fileid`=`fc`.`parent` WHERE `fc`.`fileid` = ?

-- on a delete hook remove the file or all files of a share from the shares_files relation
-- when deleting a file (or folder)
DELETE FROM `oc_shares_files` WHERE `fileid` IN (SELECT `fileid` FROM `oc_filecache` WHERE `path` LIKE ? )
-- when deleting a share
DELETE FROM `oc_shares_files` WHERE `shareid` = ?
```

If we were using foreign keys we could even use the cascade option to remove entries from the relation table.

Migration to a relations table is easy because it only adds the relation table.

It is basically a speed tradeoff where we maintain the relations of the shared files in the database to speed up all queries related to file shares, because we can let the database handle permissions and don't have to go through php and the connection.

Open questions pointed out by @schiesbn:
What happens when a different storage is mounted inside the shared folder?
My initial thought is that, since the whole storage is shared, it makes sens to have a `oc_shares_storages` table to track this case.

Let me take a step back and think out loud here. Our `oc_filecache` should contain every file only once. The `oc_storages` is basically used to grant access to a subset of these files to users. The `oc_share` table actually does the same. The storages table also serves another purpose: naming the _physical_ mount points in the _virtual_ directory tree. By creating a share we effectively create a new storage that contains a subtree of the _virtual_ directory tree. This subtree can contain any number of storages: _physical_ mount poins as well as other _virtual_ storages. ... mind blown ... Mathemathically speaking we are dealing with sets. Sets of existing files (storages) and sets of accessible files which is a subset of the filecache. The root for each accessible set is either the root of a storage or a share entry. The diffeerence is that with storages we have a way of joining or intersecting the set of files (because they share the same storage id) whereas with shares we cannot use joins or intersections, because they only contain the root element.

All of this gets even more interesting when a folder has been shared twice to the same user (via a->b, a->c, b->d, c->d) 

It seems to me we are hiding access controls in storages and shares. Maybe we should make them explicit. I need to give this some more thought ... input is welcome, maybe from someone who is much more familiar with set theory than I am.

**Comment by oparoz** [comment]
It's crucial for searchByMime() to have a limit argument so that apps which need to collect references to files of a certain type can do it in batch instead of having to send an array containing potentially thousands of entries to the browser.

**Comment by oparoz** [comment]
getDirectoryListing() is also affected, but it will probably die once search supports a path and limits.

**Comment by jospoortvliet** [comment]
@DeepDiver1975 just curious if this got on your radar at some point (or shouldn't be) ;-)

**Comment by LukasReschke** [comment]
This makes some stuff _really_ cumbersome to use, just a decent example:

![2015-10-20_11-46-30](https://cloud.githubusercontent.com/assets/878997/10603872/f35066fe-7720-11e5-97f7-83550cb6dd78.png)

@cmonteroluque Please re-evaluate whether this really should be backlog. This makes for example the gallery pretty much unusable on S3. Try to open the gallery on S3 (especially the "Shared" folder) and you will see what I mean :wink:

**Comment by LukasReschke** [comment]
Ok. I think this is caused by file locking: https://blackfire.io/profiles/f1d6349f-4ab3-4190-beab-ccbd59527844/graph

@icewind1991 Ideas?

**Comment by LukasReschke** [comment]
Moving that one out to https://github.com/owncloud/core/issues/19888

**Comment by cdamken** [comment]
@MorrisJobke @butonic 

00004526

**Comment by cdamken** [comment]
@felixboehm @bboule @butonic Is there any change to fix in a earlier version than 9.0?

**Comment by bboule** [comment]
@cdamken can you email the link to the SFDC ticket associated with this?

**Comment by cdamken** [comment]
Sure! done!

**Comment by DeepDiver1975** [comment]
Alternative and/or additional tables to implement the file cache and/or sharing is definitly out of scope for 9.0. We need an architectural discussion upon this -> 9.1 candidate.

@cmonteroluque @karlitschek

**Comment by MorrisJobke** [comment]
> 00004526

This is a case where each search on shared files took around 1 minute. As this is still in planning phase it's unlikely that this will be in 9.0, or am I wrong? @cdamken is just asking for a quick fix/work around.

cc @MTRichards @karlitschek @cmonteroluque

**Comment by bboule** [comment]
thanks @MorrisJobke is there a quick workaround that can be provided?  is this possible or do we need to wait?

**Comment by MorrisJobke** [comment]
> thanks @MorrisJobke is there a quick workaround that can be provided? is this possible or do we need to wait?

There is no real work around. Searches just take that long and a proper cache needs to be implemented. It is simply an enhancement. :(

**Comment by felixboehm** [comment]
30 sec for searching is not a normal time. Same search on same files takes only 1 sec if done by the file owner instead of users, who got the files as share.

So owncloud search not useable with shares?

**Comment by cmonteroluque** [comment]
Thinking whether @icewind1991's proposal also may help us refresh external storage changes with a more efficient file scan. Either way, it seems that we don't have an agreed solution yet. So, with @DeepDiver1975's comment, can this wait to 9.1? @karlitschek @PVince81 @bboule @MTRichards

**Comment by PVince81** [comment]
@icewind1991 has there been any improvement in 9.0 regarding search inside mount points ? I remember some changes in the storage/cache front, but not sure if it would help.

**Comment by karlitschek** [comment]
I think the proposed approach has potential. But as @DeepDiver1975 said, this is a change that needs some planing and discussion. Not sure we can/should try to squeeze this into 9.0

**Comment by bboule** [comment]
@cdamken @felixboehm given there is no quick fix for this... from a blue ticket perspective can we message this to the user (i.e. rough time frames etc.).

**Comment by felixboehm** [comment]
No, there is no quick fix or workaround for now.

**Comment by felixboehm** [comment]
@butonic would the full text search app be a workaround?

**Comment by icewind1991** [comment]
There have been changes in the 9.0 implementation of the shared cache search methods, haven't done any performance measurements though

**Comment by butonic** [comment]
only the elastic search based one for ee because it can search in shared files. but it has never been in production AFAIK. so needs an update and thorough testing.

**Comment by felixboehm** [comment]
using the search elastic app now.

**Comment by felixboehm** [comment]
@butonic close this?

**Comment by PVince81** [comment]
Already moved twice, no feedback, am closing this now. Please reopen if important.

**Comment by butonic** [comment]
Still a problem without search elastic. search lucene could be used to provide a solution for small installations. would need some work, https://github.com/owncloud/search_lucene/issues/10.

or we could add this as another performance test via smashbox, cc @mrow4a

**Comment by PVince81** [comment]
@butonic any estimate / breakdown of tasks to improve this ?

So far it feels like this is a rather big project, moving to backlog.

**Comment by PVince81** [comment]
@cdamken FYI ^
